src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach(action => {
+    const nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        for (const key in nextState) {
+          delete nextState[key];
+        }
+        break;
+      case 'addProperties':
+        Object.assign(nextState, action.extraData);
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach(key => delete nextState[key]);
+        break;
+      default:
+        break;
+    }
+
+    stateHistory.push(nextState);
```

Comment:
The nextState variable won't be assigned a valid value, so it will be undefined, and the push statement will add undefined to the array. Thanks for the heads up! I'm having an awesome learning experience here.